### PR TITLE
many: add support for downloading components to store tooling

### PR DESCRIPTION
**Pull request by alfonsosanchezbeato**
This has been split out from https://github.com/snapcore/snapd/pull/14181.

### CODE REVIEW

File: store/tooling/tooling.go [hunk 1]
```diff
@@ -221,7 +231,61 @@ func (tsto *ToolingStore) DownloadSnap(name string, opts DownloadSnapOptions) (d
...
+			return nil, fmt.Errorf("%s component for %s not found in store",
+				comp, sar.SnapName())
+		}
+		baseName := opts.Basename
+		if baseName == "" {
+			cpi := snap.MinimalComponentContainerPlaceInfo(
+				srr.Name, snap.R(srr.Revision), sar.SnapName())
+			baseName = cpi.Filename()
+		} else {
+			baseName += fmt.Sprintf("+%s.comp", srr.Name)
```
Comment (pedronis) [inline store/tooling/tooling.go:277]: this case doesn't seem to be tested but it seems the same logic is not tested for snaps either

File: store/tooling/tooling.go [hunk 2]
```diff
@@ -221,7 +231,61 @@ func (tsto *ToolingStore) DownloadSnap(name string, opts DownloadSnapOptions) (d
...
+func (tsto *ToolingStore) downloadComponents(comps []string, sar *store.SnapActionResult, opts DownloadSnapOptions) ([]*DownloadedComponent, error) {
+	downloadedComps := make([]*DownloadedComponent, len(comps))
+	for i, comp := range comps {
+		var srr *store.SnapResourceResult
+		for _, res := range sar.Resources {
+			if res.Name == comp {
+				srr = &res
+				break
+			}
+		}
```
Comment (pedronis) [inline store/tooling/tooling.go:266]: nitpick: I wonder if this would make sense as a method on sar ?

File: store/tooling/tooling.go [hunk 3]
```diff
@@ -437,3 +531,64 @@ func (tsto *ToolingStore) AssertionMaxFormats() map[string]int {
...
+	}()
+
+	dlOpts := &store.DownloadOptions{LeavePartialOnError: opts.LeavePartialOnError}
+	cref := naming.NewComponentRef(snapName, srr.Name)
+	if err = tsto.sto.Download(context.TODO(), cref.String(), targetFn,
+		&srr.DownloadInfo, pb, nil, dlOpts); err != nil {
+		return nil, err
+	}
+
+	signal.Reset(syscall.SIGINT)
```
Comment (pedronis) [inline store/tooling/tooling.go:580]: nitpick: wonder if we should extract  the ProbressBar and Signal logic that is the same for snaps in a small helper that takes a closure/function ?

**Review by pedronis — APPROVED** [review]
thanks, couple of nitpicks and testing comment

File: store/tooling/tooling.go (hunk 3 above)
Comment (alfonsosanchezbeato) [reply store/tooling/tooling.go:580]: Good idea, I've done this now

**Review by alfonsosanchezbeato — COMMENTED** [review]
@pedronis thanks for the comments. I've addressed them and rebased to master to get the latest spread test fixes.

**Review by andrewphelpsj — APPROVED** [review]
Thanks!

btw, conflicts with master now.

**Pull request merged**